feat(config): migrate wallet to use a layered configuration

gustavovalverde opened the pull request:
### Motivation

This pull request migrates Zallet's configuration system from Abscissa to `config-rs`, to unify the configuration approach across the Z3 stack (Zebra, Zaino, and Zallet).

This work follows the same pattern used in Zebra's configuration migration ([Zebra PR #9768](https://github.com/ZcashFoundation/zebra/pull/9768)) and Zaino's migration ([Zaino PR #469](https://github.com/zingolabs/zaino/pull/469)).

### Solution

The new implementation uses `config-rs` to load configuration from three sources in order of precedence:

1. Hard-coded defaults (lowest precedence)
2. Optional `zallet.toml` configuration file 
3. Environment variables with `ZALLET_` prefix (highest precedence)

Key technical improvements:

- **Native error handling**: Uses `config::ConfigError` directly, providing structured error information with file paths, keys, and type mismatches
- **Nested environment variables**: Supports `ZALLET_SECTION__KEY` format for complex configuration structures
- **Array parsing**: Comma-separated lists for arrays (e.g., `ZALLET_RPC__BIND=127.0.0.1:8080,127.0.0.1:8081`)

> The Abscissa `Configurable` trait is retained for application lifecycle integration, CLI argument injection (like `--datadir`), and command-specific configuration overrides, but no longer handles configuration file loading.

### Implementation Details

- Added `#[serde(default)]` to all nested configuration sections to ensure partial configurations work correctly
- Configuration loading logic moved to `config.rs` for better organization
- Error conversion from `config::ConfigError` to `FrameworkError` happens only in `process_config()` 
- CLI arguments like `--datadir` and `--config` are properly integrated with the configuration system

### Tests

A test suite has been added at `wallet/zallet/tests/config.rs` covering:
- Default value loading and TOML file parsing
- Environment variable overrides with nested structures
- Configuration source precedence (env > file > defaults)
- Error handling for invalid TOML, missing files, and type mismatches
- Array parsing from environment variables (both comma-separated and JSON formats)

### Specifications & References
- [`config-rs` crate documentation](https://docs.rs/config/latest/config/)

gustavovalverde committed:
feat(config): migrate wallet to use a layered configuration

This commit replaces the Abscissa-based configuration system in Zallet with
the `config-rs` crate, providing a standard layered configuration approach
that unifies the Z3 stack configuration strategy across Zebra, Zaino, and Zallet

This change introduces a clear precedence for configuration sources:
1. Hard-coded defaults (lowest precedence)
2. Optional `zallet.toml` configuration file
3. Environment variables with `ZALLET_` prefix (highest precedence)

The Abscissa `Configurable` trait remains for application lifecycle integration,
CLI argument injection, and command-specific configuration overrides.

str4d reviewed (changes requested):
I am currently on PTO, but I want to ensure this is blocked on my review.

gustavovalverde committed:
feat(config): block sensitive env var overrides with suffix-based filter

- Add is_sensitive_leaf_key() to filter ZALLET_* env vars by leaf key (case-insensitive)
- Block env overrides ending with: password, secret, token, cookie, private_key
- Pre-filter env vars before passing to config::Environment via .source(Some(filtered_env))
- Preserve existing prefix/separator logic and list parsing for rpc.bind
- Keep non-sensitive fields overridable (validator_address, validator_cookie_path, etc.)

Aligns with Zebra and Zaino security model for Z3 stack consistency.

str4d commented on zallet/src/config.rs:
```diff
@@ -35,34 +46,117 @@ pub struct ZalletConfig {
+            }
+        }
+
+        // Add filtered environment variables (highest precedence)
+        builder = builder.add_source(
```
In the group chat, we discussed only supporting env-var setting in Docker images specifically. If we take that approach, this would be where we configure it:
```suggestion
        // Add filtered environment variables (highest precedence)
        #[cfg(zallet_docker_image)]
        builder = builder.add_source(
```

str4d reviewed (changes requested):
Reviewed ca41f6e2fc36f490c26235f4381a18eaca40be58.

Please squash all undos into the first commit.

gustavovalverde committed:
chore: apply suggestions from code review

Co-authored-by: Jack Grigg <thestr4d@gmail.com>

gustavovalverde committed:
chore: apply suggestions from code review

Co-authored-by: Jack Grigg <thestr4d@gmail.com>

gustavovalverde committed:
refactor(config): address PR review comments for config-rs migration

- Move CONFIG_FILE constant from config.rs to commands.rs
- Import functions directly in config.rs instead of using fully qualified paths
- Move resolve_config_path method from ZalletConfig to EntryPoint::config_path
- Position config_path method after datadir method as requested
- Fix test compilation errors with mutable env variables as some tests failed because env variables needed mut to call env.set_var()

gustavovalverde committed:
chore: reduce diff

gustavovalverde committed:
fix(supply-chain): resolve cargo vet audit issues

- Trust Ed Page for config crate dependency (already trusted by major orgs)

gustavovalverde committed:
docs(config): document serde(default) section header behavior

Update ZalletConfig documentation to explain layered configuration approach:
- Document Option<T> field usage for default vs explicit value distinction
- Explain serde(default) section header UX balance
- Clarify config generation vs parsing behavior differences

gustavovalverde committed:
fix(config): error on sensitive environment variables instead of silently ignoring

Replace silent ignoring of sensitive environment variables with explicit errors to prevent:
- Silent misconfigurations where users think secrets are configured but aren't
- Process table exposure of sensitive values even when ignored

gustavovalverde commented on zallet/src/config.rs:
```diff
@@ -35,34 +46,117 @@ pub struct ZalletConfig {
 }
 
+/// Zallet Configuration Filename
+pub const CONFIG_FILE: &str = "zallet.toml";
```
Addressed in [`c30ceed` (#194)](https://github.com/zcash/wallet/pull/194/commits/c30ceed2a13ef39e2187a8a69483f1f45bd62eef)

gustavovalverde commented:
> Please squash all undos into the first commit.

Once we're done I'll squash. In the meanwhile I'll keep them separate for clarity

gustavovalverde committed:
chore: update `config-rs` to latest version

gustavovalverde committed:
Merge branch 'main' into feat-layered-config

gustavovalverde committed:
fix: merge conflict

arya2 commented on zallet/src/config.rs:
```diff
@@ -35,34 +46,117 @@ pub struct ZalletConfig {
+                if !is_sensitive_leaf_key(leaf_key) {
+                    filtered_env.insert(key, value);
+                }
+                // Sensitive keys are silently ignored
```
Is there an advantage to panicking if a sensitive key is detected, or is it okay to log a warning and ignore the key?
